### PR TITLE
SidePanel: Show Stashes

### LINKED ISSUE #5460: Show stashes tree in sidebar
**Do you want to request a *feature* or report a *bug*?**
feature

**What is the current behavior?**
No tags and stahes in sidebar

- GitExtensions version:
3.0
- GIT version: 
2.9
- OS version: 
Windows
- .NET version: 
4.7

Is it possible to have something similar to SourceTree tags and stashes in sidebar ?

### PR DESCRIPTION
**Pull request by gerhardol**
Fixes #5460 
Follow-up to #10138 (that added stashes to the grid)

## Proposed changes

1. Side panel: Refactor classes and names

Renamed and restructured Node/Tree classes.
Some methods were added on parents but only used in leaves
(especially for submodule vs the "revision grid" branches/tags).
Example for a moved method is aheadBehind for LocalBranchNode.

Renamed and added some classes:
* BaseRevisionTree and BaseBranchNode.cs -> BaseRevisionNode.cs
  for methods pointing to revisions in the grid (i.e. all but submodules).
* BaseRefTree for Refs (Branches/Tags).

Methods were moved to the class where they were used

2. SidePanel:  Visibility for branches and tags was not updated.

If the revision grid had another filter than current branch or
filtered branches, the branches/tags were always set as visible.

The refs are always set as visible at creation and set as
not visible after the grid has loaded.

3.  SidePanel: Show Stashes

Context menu to open/apply/pop/drop as well as stash/stash-staged/manage.

4. RemoteBranches: Show ahead/behind for remotes

Supprt Alt-Click to select the tracking local branches

This is a candidate for 4.0

### Before

image

Revision filter hiding these branches
image

### After

image

image

image

image

Remote had no ahaed/behind info (no alt-click)
image
image

## Test methodology 

Tests are adapted to additions.

## Merge strategy

Rebase merge

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/BranchTreePanel/RepoObjectsTree.cs
Comment (RussKie) [inline GitUI/BranchTreePanel/RepoObjectsTree.cs:255]: Could this be?
```suggestion
            _stashTree.Refresh(new(() => Array.Empty<GitRevision>()));
```

File: GitUI/BranchTreePanel/RepoObjectsTree.cs
Comment (RussKie) [inline GitUI/BranchTreePanel/RepoObjectsTree.cs:415]: [nit]
```suggestion
            _stashTree = new(rootNode, UICommandsSource, _refsSource);
```

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:891]:
```suggestion
            Lazy<IReadOnlyCollection<GitRevision>> getStashRevs =
                new(() => (!AppSettings.ShowStashes 
                    ? Array.Empty<GitRevision>()
                    : new RevisionReader(capturedModule, hasReflogSelector: true).GetStashes(cancellationToken)));
```

I also suspect that one set of parens is superficial.

File: GitUI/UserControls/RevisionGrid/RevisionLoadEventArgs.cs
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/RevisionLoadEventArgs.cs:7]:
```suggestion
        public RevisionLoadEventArgs(IWin32Window? ownerForm, IGitUICommands gitUICommands, Lazy<IReadOnlyList<IGitRef>> getRefs, Lazy<IReadOnlyCollection<GitRevision>> getStashRevs, bool forceRefresh)
```

**Review by RussKie — APPROVED** [review]
👍 with few nits and suggestions

**Comment by RussKie** [comment]
> image

I don't think it's improving the UI to have the same icon repeated, it's likely achieving the opposite result. We don't have the icons in the toolbar either for the same reason:
image

We either need dedicated icons, or have no icons at all.

File: GitUI/BranchTreePanel/RepoObjectsTree.cs
Comment (gerhardol) [reply GitUI/BranchTreePanel/RepoObjectsTree.cs:415]: When we talked about this feature, we have previously said that the type should be stated if not on left side of '='
These rules should be documented...
I prefer the suggestion, but do not change right now.

File: GitUI/BranchTreePanel/StashNode.cs
Comment (gerhardol) [inline GitUI/BranchTreePanel/StashNode.cs:14]: #10161 changed to display this in FormStash
image

The RevGrid show the complete ref (excluding refs/)
image

Should the list present as in FormStash?
```suggestion
            : base(tree, reflogSelector.RemovePrefix("refs/").RemovePrefix("stash"), visible)
```
image

Comment (mstv) [reply GitUI/BranchTreePanel/StashNode.cs:14]: I have no strong opinion regarding the colon - not really needed but for consistency...

```suggestion
            : base(tree, reflogSelector.RemovePrefix(GitRefName.RefsStashPrefix), visible)
```

File: GitUI/BranchTreePanel/StashTree.cs
Comment (mstv) [inline GitUI/BranchTreePanel/StashTree.cs:52]:
```suggestion
                // Stashes does not support filtering, but stashes may not be visible
```

File: GitUI/BranchTreePanel/StashTree.cs
Comment (mstv) [inline GitUI/BranchTreePanel/StashTree.cs:55]:
```suggestion
                Node? parent = node.CreateRootNode(pathToNodes, (tree, parentPath) => new BasePathNode(tree, parentPath));
```

**Review by mstv — COMMENTED** [review]
LGTM, just nits, have not run yet

File: GitUI/BranchTreePanel/StashNode.cs
Comment (gerhardol) [reply GitUI/BranchTreePanel/StashNode.cs:14]: Changed this, updated screenshot

**Review by mstv — APPROVED** [review]
Thank you, very useful

We should not attempt to select unavailable stash revisions though:

image

**Comment by gerhardol** [comment]
> We should not attempt to select unavailable stash revisions though:

You can still "open stash" after dismissing the popup.
The popup should not be raised when right clicking: #10250
This disables double clicking

**Comment by vbjay** [comment]
Check button and let through for double click of left / primary button.

**Comment by gerhardol** [comment]
Added an incomplete fix for visibility for branches, see updated description.
Normally, the grid is updated just before the branches etc were added. If non-default sorting is used, the sidepanel is updated before the grid and the visibility is incorrect.
For stashes, this was always the case and I added stashes after loading.
The handling was changed for stashes, creating the sidepanel when loading the grid but updating the visibility after the grid was loaded.
Loading earlier is not important but a similar change should be done for branches etc, but there will have to be some more changes.
This second part could be done in a sepate PR too, to get translation changes out.

A related fix:
* Visibility for branches and tags were incorrect if filters other than
branch filers were used.

File: GitUI/BranchTreePanel/StashTree.cs
Comment (mstv) [inline GitUI/BranchTreePanel/StashTree.cs:67]: ❔ `Invoke`

**Review by mstv — APPROVED** [review]

**Comment by gerhardol** [comment]
It is not obvious how to restructure the info in a good way (to not add null ObjectId to folders, may need to do that anyway or duplicate some code in tags/local/remote/stash).
There is some renaming I would like to do before merging to reflect usage better, will wait a little until I decide...

> Check button and let through for double click of left / primary button.

I guess this was related to https://github.com/gitextensions/gitextensions/pull/10258#pullrequestreview-1143339265, https://github.com/gitextensions/gitextensions/issues/10250
I tried to debug that but did not find the issue.

File: GitUI/BranchTreePanel/StashTree.cs
Comment (gerhardol) [reply GitUI/BranchTreePanel/StashTree.cs:67]: Make this an EventHandler?
What is the benefit in this scenario?

File: GitUI/BranchTreePanel/RepoObjectsTree.cs
Comment (RussKie) [reply GitUI/BranchTreePanel/RepoObjectsTree.cs:417]: We shouldn't use `var` if the type is not obvious from the expression. When newing up the type could be omitted.

This is a non-blocking suggestion.

File: GitUI/BranchTreePanel/RepoObjectsTree.cs
Comment (RussKie) [inline GitUI/BranchTreePanel/RepoObjectsTree.cs:248]: I find `RefreshWhenLoading` and `RefreshAfterLoaded` confusing. Without reading the xml-docs I couldn't quite figure out what these two do...
I'm having troubles with the meaning of "refreshing while data is loading". `RevisionGrid.RevisionsLoading` event is raised _before_ data is loaded, and `RevisionGrid.RevisionsLoaded` is raised _after_ the data has been loaded and displayed. According to the comment, we have added `RevisionGrid.RevisionsLoading` event specifically to refresh the left panel.

Was there a specific reason why you thought the two methods were necessary? Couldn't we have the single method with an additional flag to indicate whether the data _is being_ or _has been_ loaded?

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:893]: Does it have to be `new List<>`?

File: GitUI/BranchTreePanel/RepoObjectsTree.cs
Comment (RussKie) [inline GitUI/BranchTreePanel/RepoObjectsTree.cs:248]: Looks like in different parts we have different order of the same arguments:
image

Comment (gerhardol) [reply GitUI/BranchTreePanel/RepoObjectsTree.cs:248]: Before it was Refresh() in many situations which was more confusing.
The first call is before the revision grid is available. The data to add is there, but not if a revision is visible or not, which is available after loading. RefreshAfterLoaded() do not need any parameters right now, it just updates the visibility.

The previous commit updated stash info after loading only. That strategy could be used for all 'commit' (branches, tags, stashes but not submodules), but then updating of will be delayed.

So I prefer two methods, possibly with updated arguments.

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs
Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:893]: Probably not
This is some us for something done once in a process that may take several seconds, so there is not much point in optimizing.
(I have spent a lot of time optimizing also single ms for 4.0, so performance is not unimportant but focus where it makes a difference.)

Comment (mdonatas) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:893]: In this case it's not even about performance. This simply looks wrong. To something which is **ReadOnly** a collection meant for manipulation is assigned. This raises questions and wastes others time while trying to comprehend the meaning of the code.
`Array.Empty<GitRevision>()` set a good tone and communicates intent clearly.

File: GitUI/BranchTreePanel/RepoObjectsTree.cs
Comment (gerhardol) [reply GitUI/BranchTreePanel/RepoObjectsTree.cs:248]: Renaming to `RefreshRevisionsLoading` and `RefreshRevisionsLoaded` so similar to the events.
Keeping the arguments the same, as isFiltering is needed after loading (to skip updating visibility).

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs
Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:893]: Updating
(I do not find Array.Empty to look better in any way.)

**Comment by gerhardol** [comment]
Rebased and added three commits:
* review comments
* refactoring without functional changes
* Visibility for branches and tags were incorrect if filters other than branch filers were used.

```
SidePanel: Show Stashes

Context menu to open/apply/pop/drop.

Renamed and restructured Node/Tree classes:
* BaseRevisionTree and BaseBranchNode.cs -> BaseRevisionNode.cs
  for methods pointing to revisions in the grid (i.e. all but submodules).
* BaseRefTree for Refs (Branches/Tags).

Methods were moved to the class where they were used
(like aheadBehind in the LocalBranchNode).

Some renaming to clarify use

* Visibility for branches and tags were incorrect if filters other than
branch filers were used.
```

The Latest refactoring changes some that was added in the first commit and I would like to include it.
I could rebase to make the refactoring first, then 'filter' visibility fixes, then Stash.

There is one omission in features too:

image
image

File: GitUI/BranchTreePanel/RepoObjectsTree.cs
Comment (RussKie) [reply GitUI/BranchTreePanel/RepoObjectsTree.cs]: > I will not name who set the original order in the event handlers, the side panel methods or who requested changes previously in this PR. We all have a right to change our minds.

😆

File: GitUI/BranchTreePanel/StashTree.cs
Comment (mstv) [reply GitUI/BranchTreePanel/StashTree.cs]: I think `Invoke` is the short of `RunAsync(await SwitchToMainThread(); actual call;)`. But I don't mind using JTF for all async operations.

File: GitUI/BranchTreePanel/LocalBranchNode.cs
Comment (mstv) [inline GitUI/BranchTreePanel/LocalBranchNode.cs]: This used to work for remote branches, too...

File: GitCommands/Settings/AppSettings.cs
Comment (mstv) [inline GitCommands/Settings/AppSettings.cs:1925]: I have moved it to the top.
(The weird auto-reload of settings without UI update confused me once again. If another GE instance denies to work like for `Move up`, hit `F5`.) 
The `Stashes` tree is collapsed by default. So it does not consume much space but is at hand quickly.
(not requesting a change)

**Review by mstv — COMMENTED** [review]
Clear refactoring 👍

I like feature/rot-stashes-2.

File: GitCommands/Settings/AppSettings.cs
Comment (gerhardol) [reply GitCommands/Settings/AppSettings.cs:1925]: The index is stored in settings, so it is safer to store new occurrences last.
Do you request a change?

File: GitUI/BranchTreePanel/LocalBranchNode.cs
Comment (gerhardol) [reply GitUI/BranchTreePanel/LocalBranchNode.cs]: Related branch was not set only for Local branch. I was thinking about this myself.
Not tonight, probably separate PR

**Comment by gerhardol** [comment]
Three more commits with one fix, one cleanup in the optimization and addition of context menu on the root stash (screenshot updated).

> Clear refactoring
I like feature/rot-stashes-2.

Squashed and refactored into new commits with refactoring, visibility for branches etc at filtering and stash related separated.
This is how I would like to merge this.

File: GitCommands/Settings/AppSettings.cs
Comment (mstv) [reply GitCommands/Settings/AppSettings.cs:1925]: >> (not requesting a change)
>
> Do you request a change?

I just wanted to raise this point. Perhaps others vote for a change.

> The index is stored in settings, so it is safer to store new occurrences last.

Yes, else it might become difficult to handle.

File: GitUI/BranchTreePanel/LocalBranchNode.cs
Comment (mstv) [reply GitUI/BranchTreePanel/LocalBranchNode.cs]: OK

Comment (gerhardol) [reply GitUI/BranchTreePanel/LocalBranchNode.cs]: I doubt that this has ever worked.
The solution should also eliminate the additional Git command needed for push button vs left panel (as remote/local should share the info anyway). The aheadBehindProvider should be made lazy or the info be provided as the refs. Lazy data reset when the refs are refreshed.

Slightly offtopic, but should remote branches be decorated in some way?
The easiest is to have exactly the same info as locals. The info could be prepended with "local: " or so but that is annoying when you know about it. Tooltip could add details.

File: GitCommands/Git/AheadBehindDataProvider.cs
Comment (RussKie) [inline GitCommands/Git/AheadBehindDataProvider.cs:12]:
```suggestion
        void Initialize();
```

Whether it's lazy or not is an implementation detail.

File: GitCommands/Git/AheadBehindDataProvider.cs
Comment (RussKie) [inline GitCommands/Git/AheadBehindDataProvider.cs:52]: Is this needed?

File: GitUI/BranchTreePanel/BaseBranchLeafNode.cs
Comment (RussKie) [inline GitUI/BranchTreePanel/BaseBranchLeafNode.cs:72]: It should be possible to write it as:
```suggestion
            TreeViewNode.TreeView?.BeginInvoke(() =>
```

File: GitCommands/Git/AheadBehindDataProvider.cs
Comment (gerhardol) [reply GitCommands/Git/AheadBehindDataProvider.cs:12]: ResetCache() or just Reset may be better then - it is resetting the cached result.

File: GitCommands/Git/AheadBehindDataProvider.cs
Comment (gerhardol) [reply GitCommands/Git/AheadBehindDataProvider.cs:52]: The commented out code is to document the behavor. I would have liked to enable the Debug if that had been working in .NET > 4.
For now it auto-resets the cache to get results for the branches.
Should just occur in tests.

File: GitUI/BranchTreePanel/BaseBranchLeafNode.cs
Comment (gerhardol) [reply GitUI/BranchTreePanel/BaseBranchLeafNode.cs:72]: Will change
This was just moved from BaseBranchNode.cs
I assume this goes into the "refactoring - no changes" commit

**Comment by gerhardol** [comment]
Pushed some more refactoring - had not planned to do that in this PR, but pushing it here to avoid changing the same class twice.
Some duplicate code i inherited functions could be removed.

The actual change for https://github.com/gitextensions/gitextensions/pull/10258#discussion_r999961425
was then quite small and was included here.

feature/rot-stashes-2 is updated

I still want to include this in 4.0

**Comment by RussKie** [comment]
I'd like to get v4 out of the door. We keep dragging on and on with that...

**Comment by gerhardol** [comment]
> I'd like to get v4 out of the door. We keep dragging on and on with that...

Agree in general, but this 'completes' a feature

**Comment by RussKie** [comment]
OK. Let's get it merged then

File: GitUI/BranchTreePanel/RepoObjectsTree.cs
Comment (gerhardol) [reply GitUI/BranchTreePanel/RepoObjectsTree.cs]: I renamed the repoObjectsTree method `Refresh` to `RevisionGrid.RevisionsLoading -> RefreshRevisionsLoading` and added `RevisionGrid.RevisionsLoaded -> RefreshRevisionsLoaded`, called from .
The call to RepoObjectTree is now wrapped in `FormBrowse.RefreshLeftPanel`, but keeping the changed names.

**Comment by gerhardol** [comment]
Need to rebased on master,  #10283
I would like to get a review from @mstv too.
The cache handling in part 4 (that saves a Git call ) can be done in other way, I believe this OK

**Comment by gerhardol** [comment]
Squashed to descriptive commits, rebased on master and #10283. Especially #10285 required some redesign.
Some other refining too.

See changes also due to other changes in `feature/rot-stashes-2` that now contain the old PR branch.

**Comment by gerhardol** [comment]
Plan to rebase merge this tomorrow, unless there are review comments

File: GitUI/BranchTreePanel/StashTree.cs
Comment (mstv) [inline GitUI/BranchTreePanel/StashTree.cs:39]: ❔ Do it like in BaseRevisionTree?
```
    <snip>
}).FileAndForget();
```
---

General question: Might `ConfigureAwait()` be the cause why the app sometimes crashes without any popup?

**Review by mstv — COMMENTED** [review]
Just reviewed on GH

File: GitCommands/Git/AheadBehindDataProvider.cs
Comment (gerhardol) [reply GitCommands/Git/AheadBehindDataProvider.cs:58]: It is a way of syncing.
Something like `_data ??= GetData(null, branchName);` may cause GetData to be called three times, but using Lazy here blocks this.
Could be used with a lock() or so instead.
Can add a comment.

File: GitUI/BranchTreePanel/BaseRefTree.cs
Comment (gerhardol) [reply GitUI/BranchTreePanel/BaseRefTree.cs:99]: Reloading is always done before UpdateVisibility(), the visibility for refs is always set to true.

File: GitUI/BranchTreePanel/StashTree.cs
Comment (gerhardol) [reply GitUI/BranchTreePanel/StashTree.cs:39]: Same for Branches/Tags in BaseRefTree.cs, so not changed.
I guess it could be changed.

> General question: Might ConfigureAwait() be the cause why the app sometimes crashes without any popup?

Maybe Debug.Assert or Debug.Fail that fires if you run in Debug. Hard to debug. Maybe fixed in .net7
https://github.com/dotnet/winforms/issues/3739

**Comment by pmiossec** [comment]
Not a critic but more a remark...

I had to think some longs seconds before understanding why some recent stashes are greyed:
image

before understanding that are the ones based on a deleted commit only reachable when displaying the reflog.

Even if it is right, it is not straightforward and I think we could expect some questions on that after the release 😉 (maybe this comment will be a good answer for those used to search before opening an issue...)

**Comment by gerhardol** [comment]
> I had to think some longs seconds before understanding why some recent stashes are greyed: 

There could be a tooltip to enable reflog to see them in the grid, or doubleclick to open the stash viewer (that has improved navigation in 4.0 too).

I assume you expect all to be seen with _All_ branches, branches not visible when branches are filtered have the same decoration.

**Pull request merged**
